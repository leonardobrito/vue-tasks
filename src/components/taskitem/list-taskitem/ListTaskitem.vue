<template>
  <div>
    <label>{{ taskItemsOverview }}</label>
    <br />
    <ul>
      <li v-for="taskItem in formattedTaskItems" :key="taskItem.id">
        <div>
          <label>checkbox:</label>
          <input
            type="checkbox"
            v-model="taskItem.checked"
            @click="updateCheckTaskItem(taskItem)"
          />
        </div>
        <label>name: {{ taskItem.name }}</label>
        <br />
        <label>description: {{ taskItem.description }}</label>
        <br />
        <button @click="updateTaskitem(taskItem.id)">Update item</button>
        <br />
        <button @click="deleteTaskitem(taskItem.id)">Delete item</button>
      </li>
    </ul>
  </div>
</template>

<script>
import { updateTaskItemApi, deleteTaskitemApi } from "../../../services/api";
import { mapState, mapMutations } from "vuex";

export default {
  methods: {
    ...mapMutations('taskItem', ['setTaskitem', 'removeTaskitem']),
    formatTaskItem(taskItem) {
      return {
        id: taskItem.id,
        checked: taskItem.checked,
        name: taskItem.task.name,
        description: taskItem.task.description
      };
    },
    isTaskItemChecked({ checked }) {
      return checked;
    },
    updateCheckTaskItem(taskItem) {
      const mutableTaskItem = {
        id: taskItem.id,
        checked: !taskItem.checked
      };
      updateTaskItemApi(mutableTaskItem).then(
        result => this.setTaskitem(mutableTaskItem),
        error => console.log(error)
      );
    },
    updateTaskitem(id) {
      this.$router.push({ name: "showTaskitem", params: { id } });
    },
    deleteTaskitem(taskitemId) {
      deleteTaskitemApi(taskitemId).then(
        result => this.removeTaskitem(taskitemId),
        error => console.log(error)
      );
      // this.removeTaskitem(taskitemId)
    }
  },
  computed: {
    ...mapState('taskItem', ['taskItems']),
    formattedTaskItems: function() {
      const { taskItems, formatTaskItem } = this;

      if (!taskItems) {
        return [];
      }

      return taskItems.map(formatTaskItem);
    },
    taskItemsOverview: function() {
      const { formattedTaskItems, isTaskItemChecked } = this;

      const length = formattedTaskItems.length;
      const checkedLength = formattedTaskItems.filter(isTaskItemChecked).length;

      return `${checkedLength} of ${length} are checked`;
    }
  }
};
</script>

<style></style>
