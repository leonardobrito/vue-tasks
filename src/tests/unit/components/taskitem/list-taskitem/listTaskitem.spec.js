import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import ListTaskitem from '@/components/taskItem/list-taskitem/ListTaskitem.vue';

const localVue = createLocalVue();
localVue.use(Vuex);

const setTaskitem = jest.fn();
const taskItem = {
  namespaced: true,
  state: {
    taskItems: [
      {
        id: 1,
        checked: false,
        task: {
          id: 1,
          name: 'Learn Vuejs',
          description: 'Description',
          user_id: 1
        }
      },
      {
        id: 2,
        checked: true,
        task: {
          id: 2,
          name: 'Do tests',
          description: 'Description 2',
          user_id: 1
        }
      }
    ],
  },
  actions: {
    setTaskitem
  },
}

describe('Components > TaskItem > ListTaskitem', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(ListTaskitem, {
      localVue,
      store: new Vuex.Store({
        modules: { taskItem }
      })
    });
  });

  it('should matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render taskItemsOverview expected text', () => {
    const expected = '1 of 2 are checked';
    expect(wrapper.vm.taskItemsOverview).toBe(expected);
  });

  describe('when taskitem is checked', () => {
    it('should call setTaskitem action', () => {
      let checkbox = wrapper.find('input');
      checkbox.trigger('click');

      expect(setTaskitem).toBeCalled();
    });
  });
});
