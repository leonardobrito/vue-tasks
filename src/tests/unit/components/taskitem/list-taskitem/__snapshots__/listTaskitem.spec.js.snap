// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components > TaskItem > ListTaskitem should matches snapshot 1`] = `
<div>
  <label>
    1 of 2 are checked
  </label>
   
  <br />
   
  <ul>
    <li>
      <div>
        <label>
          checkbox:
        </label>
         
        <input
          type="checkbox"
        />
      </div>
       
      <label>
        name: Learn Vuejs
      </label>
       
      <br />
       
      <label>
        description: Description
      </label>
       
      <br />
       
      <button>
        Update item
      </button>
       
      <br />
       
      <button>
        Delete item
      </button>
    </li>
    <li>
      <div>
        <label>
          checkbox:
        </label>
         
        <input
          type="checkbox"
        />
      </div>
       
      <label>
        name: Do tests
      </label>
       
      <br />
       
      <label>
        description: Description 2
      </label>
       
      <br />
       
      <button>
        Update item
      </button>
       
      <br />
       
      <button>
        Delete item
      </button>
    </li>
  </ul>
</div>
`;
